pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -234,204 +234,283 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, ProtocolFeeCache {
+            uint256 amountInWithFee = _addSwapFeeAmount(
+                kind == IVault.SwapKind.GIVEN_IN ? givenAmount : calculatedAmount
+            );
+            uint256 amountOut = kind == IVault.SwapKind.GIVEN_IN ? calculatedAmount : givenAmount;
 
-                uint256 amountInWithFee = _addSwapFeeAmount(request.amount);
-                balances[newIndexIn] = balances[newIndexIn].add(amountInWithFee);
-                balances[newIndexOut] = balances[newIndexOut].sub(amountOut);
+            balances[indexIn] = balances[indexIn].add(amountInWithFee);
+            balances[indexOut] = balances[indexOut].sub(amountOut);
 
-                _payDueProtocolFeeByInvariantIncrement(
-                    invariant,
-                    currentAmp,
-                    balances,
-                    virtualSupply,
-                    protocolSwapFeePercentage
-                );
-            }
+            _payDueProtocolFeeByInvariantIncrement(
+                invariant,
+                currentAmp,
+                balances,
+                virtualSupply,
+                protocolSwapFeePercentage
+            );
         }
     }
 
-    function _onSwapGivenOut(
-        SwapRequest memory request,
-        uint256[] memory balancesIncludingBpt,
+    function _onSwapWithBpt(
+        SwapRequest memory swapRequest,
+        uint256[] memory balances,
         uint256 indexIn,
-        uint256 indexOut
-    ) internal virtual override whenNotPaused returns (uint256 amountIn) {
-        _cacheTokenRatesIfNecessary();
+        uint256 indexOut,
+        uint256[] memory scalingFactors
+    ) private returns (uint256) {
+        _upscaleArray(balances, scalingFactors);
 
+        (uint256 virtualSupply, uint256[] memory balancesWithoutBpt) = _dropBptItemFromBalances(balances);
         uint256 protocolSwapFeePercentage = getProtocolSwapFeePercentageCache();
+        (uint256 amp, ) = _getAmplificationParameter();
 
-        // Compute virtual BPT supply and token balances (sans BPT).
-        (uint256 virtualSupply, uint256[] memory balances) = _dropBptItemFromBalances(balancesIncludingBpt);
-
-        if (request.tokenIn == IERC20(this)) {
-            amountIn = _onSwapBptGivenTokenOut(request.amount, _skipBptIndex(indexOut), virtualSupply, balances);
-
-            if (protocolSwapFeePercentage > 0) {
-                _payDueProtocolFeeByBpt(amountIn, protocolSwapFeePercentage);
-            }
-        } else if (request.tokenOut == IERC20(this)) {
-            amountIn = _onSwapTokenGivenBptOut(request.amount, _skipBptIndex(indexIn), virtualSupply, balances);
-
-            // For given out swaps, request.amount holds the amount out.
-            if (protocolSwapFeePercentage > 0) {
-                _payDueProtocolFeeByBpt(request.amount, protocolSwapFeePercentage);
-            }
+        bool bptIsTokenIn = swapRequest.tokenIn == IERC20(this);
+
+        // Would like to use a function selector here, but it causes stack issues
```

Comment:
What do you mean by function selector?

Reply:
Facu's trick: (boolVar ? funcA : funcB)(arg1, arg2, arg3)

Reply:
Ohh you mean function pointers. Yes, that'd be handy

Reply:
Well now I can't anyway because I'm also doing the scaling up here.